shardingsphere-kernel/shardingsphere-data-pipeline/shardingsphere-data-pipeline-core/src/main/java/org/apache/shardingsphere/data/pipeline/core/util/MigrationDistributedCountDownLatch.java

```diff
@@ -0,0 +1,149 @@
+    
+    private final Map<String, InnerCountDownLatchHolder> countDownLatchMap = new ConcurrentHashMap<>();
+    
+    private MigrationDistributedCountDownLatch() {
+        clusterPersistRepository = (ClusterPersistRepository) PipelineContext.getContextManager().getMetaDataContexts().getPersistService().getRepository();
+    }
+    
+    /**
+     * Get instance.
+     *
+     * @return instance
+     */
+    public static MigrationDistributedCountDownLatch getInstance() {
+        return INSTANCE;
+    }
+    
+    /**
+     * Register count down latch.
+     *
+     * @param parentPath parent path
+     * @param totalCount total count
+     */
+    public void register(final String parentPath, final int totalCount) {
+        clusterPersistRepository.persist(parentPath, "");
+        countDownLatchMap.computeIfAbsent(parentPath, k -> new InnerCountDownLatchHolder(totalCount, new CountDownLatch(1)));
+    }
+    
+    /**
+     * Persist ephemeral children node.
+     *
+     * @param parentPath parent path
+     * @param shardingItem sharding item
+     */
+    public void persistEphemeralChildrenNode(final String parentPath, final int shardingItem) {
+        String key = String.join("/", parentPath, Integer.toString(shardingItem));
+        clusterPersistRepository.delete(key);
+        clusterPersistRepository.persistEphemeral(key, "");
+    }
+    
+    /**
+     * Persist ephemeral children node.
+     *
+     * @param parentPath parent path
+     */
+    public void removeParentNode(final String parentPath) {
+        clusterPersistRepository.delete(String.join("/", parentPath));
+        countDownLatchMap.remove(parentPath);
+    }
+    
+    /**
+     * Await unitl all children node is ready.
+     *
+     * @param parentPath parent path
+     * @param timeout timeout
+     * @param timeUnit time unit
+     * @return
+     * true if the count reached zero and false if the waiting time elapsed before the count reached zero
+     */
+    public boolean await(final String parentPath, final long timeout, final TimeUnit timeUnit) {
+        InnerCountDownLatchHolder holder = countDownLatchMap.get(parentPath);
+        if (holder == null) {
+            return false;
+        }
+        try {
+            boolean awaitResult = holder.getCountDownLatch().await(timeout, timeUnit);
+            if (!awaitResult) {
+                log.info("await timeout, parent path: {}, timeout: {}, time unit: {}", parentPath, timeout, timeUnit);
+            }
+            return awaitResult;
+        } catch (final InterruptedException ignored) {
+        }
+        return false;
+    }
+    
+    /**
+     * Check child node count equal shardingCount.
+     *
+     * @param event event
+     */
+    public void checkChildrenNodeCount(final DataChangedEvent event) {
+        if (StringUtils.isBlank(event.getKey())) {
+            return;
+        }
+        if (!BARRIER_MATCH_PATTERN.matcher(event.getKey()).matches()) {
+            return;
+        }
```

Comment:
BARRIER_MATCH_PATTERN exists in job executor, could we remove it here? It need to maintain 2 copy of pattern, it's error prone.